Redacted events: Use Any*Event enums for redacted events

DevinR528 opened the pull request:
Sorry if opening all these is annoying, implementing it even if it's a bit off is the best way for me to understand something and I'm hoping having something to see and talk about helps everyone.

commit 1: Add most of the logic to generate the correct AnyRedactedEvent enums.
commit 2: Add the needed event content types for the 5 state events that have a content field. Create full, stub, and stripped versions of state events and full and stub versions of message event structs. Uncomment and fix redacted event tests.

farodin91 commented:
You need to update cargo first of all to travis building this.
I'll look into this later today.

jimmycuadra commented:
I managed to figure this out by just removing a bunch of generic code and using macros to reduce repetition in implementing conversions. I have one failing test which I still need to figure out. Once there is a nightly Rust without that Cargo TLS bug, I will push up my changes.

jimmycuadra commented:
Merged in c8a6832.

DevinR528 committed:
Implement redacted events using the Any*Event enums

DevinR528 committed:
Generate redacted event enums and implement corresponding event structs

Add the needed event content types for the 5 state events that have a
content field. Create full, stub and stripped versions of state events
and full and stub versions of message event structs. Uncomment and fix
redacted event tests.

DevinR528 committed:
Rename redacted enums expansion and group functions in event_enum

jplatte commented:
> implementing it even if it's a bit off is the best way for me to understand something and I'm hoping having something to see and talk about helps everyone

It definitely helps, no worries!

jplatte commented on ruma-events-macros/src/event_enum.rs:
```diff
@@ -97,10 +105,58 @@ pub fn expand_stripped_enum(input: &EventEnumInput) -> syn::Result<TokenStream>
+/// Generates the 3 redacted state enums, 2 redacted message enums,
+/// and `Deserialize` implementations.
+///
+/// No content enums are generated since only five state variants contain content.
```
And more importantly, since there's just no part of the API that deals with "any redacted event's content".

jplatte commented on ruma-events-macros/src/event_enum.rs:
```diff
@@ -97,10 +105,58 @@ pub fn expand_stripped_enum(input: &EventEnumInput) -> syn::Result<TokenStream>
+pub fn expand_content_enum(input: &EventEnumInput, ident: &Ident) -> syn::Result<TokenStream> {
     let attrs = &input.attrs;
-    let ident = Ident::new(&format!("{}Content", input.name.to_string()), input.name.span());
+    let ident = Ident::new(&format!("{}Content", ident), ident.span());
```
There's https://docs.rs/quote/1.0.7/quote/macro.format_ident.html for this kind of thing. Maybe you can switch all of these to it in a separate PR.

DevinR528 replied:
That would definitely clean things up here. I'll open one soon.

jplatte commented on ruma-events-macros/src/event_enum.rs:
```diff
@@ -401,6 +510,12 @@ fn generate_accessor(
+/// Returns true if the `ident` is a state or message event.
+fn needs_redacted(ident: &Ident) -> bool {
+    let ident = ident.to_string();
+    ident.contains("State") || ident.contains("Message")
```
I thought we had constants that the idents can be checked against now?

jplatte reviewed (commented):
First impressions are positive ^^

DevinR528 committed:
Enable the *EventContent derives to generate redacted events

Add two ruma_event attributes to optionally emit the redacted event
code. Conditionally generate a unit struct to represent the redacted
events content.

DevinR528 commented on ruma-events/src/custom.rs:
```diff
@@ -8,6 +8,40 @@ use crate::{
+    fn event_type(&self) -> &str {
+        // TODO how do we handle this?
+        "custom"
+    }
```
Should the `RedactedCustomEvent` be `RedactedCustomEvent { event_type: String }`?

jplatte replied:
Yes. Also with `#[serde(skip)]` though.

DevinR528 replied:
Without altering the `Event` derive this is what I came up with for the serialized redacted custom event

```rust
let redacted = json!({
    "content": {},
    "event_id": "$h29iv0s8:example.com",
    "origin_server_ts": 1,
    "sender": "@carl:example.com",
    "state_key": "hello there",
    "unsigned": unsigned,
    "type": "m.made.up"
});
```

The `Event` derive only skips fields if they are zst's, not sure how or if we should check if the event is custom?

jplatte replied:
I guess we need another event content trait method. `fn has_fields() -> bool` should work.

DevinR528 commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -7,42 +7,128 @@ use syn::{
+                    ) -> Result<Self, ::serde_json::Error> {
+                        // TODO error or just Ok(#redacted_ident) ??
+                        Err(::serde::de::Error::custom("redacted event content cannot be generated from a JSON value"))
+                    }
```
Any thoughts on how `from_parts` should be handled?

jplatte replied:
This should really just work the same as the other `from_parts`, except it should have only the fields with `#[skip_redaction]`. No need to specially handle ZSTs even, because `struct A;` can be instantiated as `A {}` :)

jplatte commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -7,42 +7,128 @@ use syn::{
 };
 
+mod kw {
+    syn::custom_keyword!(skip_redacted);
```
Interesting, I would have parsed to `Ident` and then checked that, but this might be the better solution actually.

jplatte replied:
This should be named skip_redact**ion** though.

jplatte commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -7,42 +7,128 @@ use syn::{
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    let event_type = {
+        let event_str = content_attr.iter().find(|a| a.is_event_type()).ok_or_else(|| {
```
If you remove `is_event_type` and add `get_event_type(&self) -> Option<&LitStr>` instead, you can use `find_map` here and don't need the `unreachable!` below.

jplatte commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -7,42 +7,128 @@ use syn::{
+                impl #ident {
+                    /// Transforms the full event content into a redacted content according to spec.
+                    pub fn redact(self) -> #redacted_ident {
+                        #redacted_ident
```
This need to handle fields that don't get removed during redaction (those with `#[skip_redaction]`).

```suggestion
                        #redacted_ident { $(…)* }
```

jplatte commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -61,65 +147,130 @@ pub fn expand_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
+        .attrs
+        .iter()
+        .flat_map(|a| a.parse_args::<EventMeta>().ok())
+        .find(|a| a == &EventMeta::SkipRedacted || a == &EventMeta::RedactedCustom)
```
Ah, you seem to have misunderstood `#[skip_redaction]`. It was supposed to be a *field* attribute, not a struct attribute, and mark fields that should stay when an event is redacted. So this really should just be

```suggestion
        .find(|a| a == &EventMeta::RedactedCustom)
```

jplatte commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -61,65 +147,130 @@ pub fn expand_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
     })
 }
+
+fn needs_redacted(input: &DeriveInput) -> bool {
```
We are already parsing the arguments elsewhere, right? Can we not give this function `&[EventMeta]` as input?

DevinR528 replied:
The problem is that all the `expand_*_event_content` functions used in `lib.rs` use the `needs_redacted` function and each of them only has access to the `input`? I could return a `Result<(TokenStream, bool)>` from `expand_event_content` maybe?

jplatte replied:
Hm, in that case let's keep this as-is, it can still be re-visited later.

jplatte commented on ruma-events-macros/src/event_enum.rs:
```diff
@@ -278,17 +287,17 @@ fn expand_any_enum_with_deserialize(
-                    let event =
-                        ::serde_json::from_str::<::ruma_events::#redacted_event>(json.get())
-                            .map_err(D::Error::custom)?;
+                    let event = ::serde_json::from_str::<
+                        ::ruma_events::#event_struct<::ruma_events::custom::RedactedCustomEventContent>,
+                    >(json.get())
+                    .map_err(D::Error::custom)?;
```
Is this how `rustfmt` would format this?

DevinR528 replied:
Yup, that one I commented out the `quote!` and ran it.

jplatte replied:
Okay. Looks kind of weird, but it's fine then.

DevinR528 replied:
I agree I dislike the hanging `>` :shrug:

DevinR528 commented on ruma-events/src/room/redaction.rs:
```diff
@@ -55,9 +55,21 @@ pub struct RedactionEventStub {
+#[derive(Clone, Debug, Deserialize, Serialize, EventContent)]
+#[ruma_event(type = "m.room.redaction")]
+#[ruma_event(custom_redacted)]
+pub struct RedactedRedactionEventContent;
```
Is this something that can happen or is it not possible to redact a redaction?

jplatte replied:
I would expect this to be possible. It seems unlikely to come up in practice, but people could create spam that's hard to remove by putting it as the reason when redacting their own messages. I doubt any clients currently expose ways to redact redactions :D

DevinR528 committed:
Use skip_redaction to denote fields kept post redaction

All redacted event code is now generated by the *EventContent derive
macro. The exception are any content structs with the custom_redaction
attribute. This leaves implementing up to the user.

DevinR528 committed:
Fix clippy useless conversion error

DevinR528 committed:
Add redact method to Redaction/CustomEventContent

DevinR528 committed:
Add accessor methods for redacted event enums

jplatte commented on ruma-events-macros/src/event_enum.rs:
```diff
@@ -345,7 +381,7 @@ fn marker_traits(ident: &Ident) -> TokenStream {
 fn accessor_methods(ident: &Ident, variants: &[Ident]) -> TokenStream {
     // We do not supply accessor methods for redacted events.
```
This comment seems to now be outdated.

jplatte commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -7,39 +7,164 @@ use syn::{
+                .cloned()
+                .collect::<Vec<_>>();
+            // remove only the field attributes that are from `ruma_events_macro` compiler
+            // fails otherwise.
```
I was briefly unclear on what this does. I think this would be clearer:

```suggestion
            // don't re-emit `ruma_events` attributes
```

WDYT?

DevinR528 replied:
Much better I was having trouble puting that into words, clearly xD

jplatte commented on ruma-events-macros/src/event_enum.rs:
```diff
@@ -164,12 +265,14 @@ fn expand_any_enum_with_deserialize(
     let attrs = &input.attrs;
     let event_type_str = &input.events;
-    let event_struct = Ident::new(&ident.to_string().trim_start_matches("Any"), ident.span());
+    let event_struct = Ident::new(&ident.to_string().replace("Any", ""), ident.span());
```
This is becoming kind of brittle. With every single `event_enum!` generating more than the specified enum after this PR, could you rewrite it to parse predefined values, sth. like `kind: state` / `kind: message` instead of the current `name: AnyStateEvent` / `name: AnyMessageEvent`, in a followup PR?

DevinR528 replied:
Yup I'll do that next.

jplatte commented on ruma-events/src/custom.rs:
```diff
@@ -8,6 +8,43 @@ use crate::{
+#[derive(Clone, Debug, Serialize)]
+pub struct RedactedCustomEventContent {
+    /// The event type string for this custom event.
+    #[serde(skip, rename = "type")]
```
The `rename = "type"` is unnecessary, as this field is never (de)serialized.

jplatte commented on ruma-events/src/enums.rs:
```diff
@@ -149,11 +172,21 @@ impl<'de> de::Deserialize<'de> for AnyRoomEvent {
+                    return Ok(AnyRoomEvent::RedactedState(from_raw_json_value(&json)?));
+                }
+            }
             Ok(AnyRoomEvent::State(from_raw_json_value(&json)?))
```
These can all be made a bit more concise:

```suggestion
            Ok(match unsigned {
                Some(unsigned) if unsigned.redacted_because.is_some() => {
                    AnyRoomEvent::RedactedState(from_raw_json_value(&json)?)
                }
                _ => AnyRoomEvent::State(from_raw_json_value(&json)?)
            })
```

jplatte commented on ruma-events/src/room/create.rs:
```diff
@@ -12,11 +12,20 @@ use crate::StateEvent;
 /// events.
 pub type CreateEvent = StateEvent<CreateEventContent>;
 
+// /// The payload for a redacted `CreateEvent`.
```
We don't need this commented-out code, right?

jplatte commented on ruma-events/src/room/history_visibility.rs:
```diff
@@ -10,9 +10,15 @@ use crate::StateEvent;
 /// The payload for `HistoryVisibilityEvent`.
 #[derive(Clone, Debug, Deserialize, Serialize, StateEventContent)]
 #[ruma_event(type = "m.room.history_visibility")]
+#[ruma_event(custom_redacted)]
```
Why `custom_redacted`?

DevinR528 replied:
These are just type aliases, this attribute skips the whole redacted code-gen.

jplatte replied:
Oh, I see. I don't think we should have them as type aliases though. I don't think it buys us much, and having redaction be 'deserialize -> .redact() -> serialize' for most types but 'deserialize -> serialize' for just a few seems weird. Plus, we want it to be a backwards-compatible change to add optional fields that don't survive redaction (although I wonder how likely it is that this will actually occur).

DevinR528 replied:
Ohh right it messes up all the methods, didn't think of that. I'll just remove that attribute for those events.

jplatte commented on ruma-events/src/room/join_rules.rs:
```diff
@@ -9,9 +9,15 @@ use crate::StateEvent;
 /// The payload for `JoinRulesEvent`.
 #[derive(Clone, Debug, Deserialize, Serialize, StateEventContent)]
 #[ruma_event(type = "m.room.join_rules")]
+#[ruma_event(custom_redacted)]
```
Same here, why `custom_redacted`?

jplatte commented on ruma-events/src/room/member.rs:
```diff
@@ -36,6 +36,15 @@ use crate::{StateEvent, StateEventStub, StrippedStateEventStub};
 /// must be assumed as leave.
 pub type MemberEvent = StateEvent<MemberEventContent>;
 
+// /// The payload for a redacted `MemberEvent`.
```
Commented out code.

jplatte commented on ruma-events/src/room/redaction.rs:
```diff
@@ -55,93 +55,33 @@ pub struct RedactionEventStub {
 /// A redaction of an event.
 #[derive(Clone, Debug, Deserialize, Serialize, EventContent)]
 #[ruma_event(type = "m.room.redaction")]
+#[ruma_event(custom_redacted)]
```
Same.

jplatte commented on ruma-events/src/room/redaction.rs:
```diff
@@ -55,93 +55,33 @@ pub struct RedactionEventStub {
 impl ruma_events::RoomEventContent for RedactionEventContent {}
 
 impl ruma_events::MessageEventContent for RedactionEventContent {}
-
```
Why did you remove these tests?

DevinR528 replied:
They were the same as in `tests/redaction.rs`

jplatte commented on ruma-events/tests/redacted.rs:
```diff
@@ -0,0 +1,270 @@
+fn redacted_deserialize_any_room() {
+    let mut unsigned = UnsignedData::default();
+    // The presence of `redacted_because` triggers the event enum (AnyRoomEvent in this case)
+    // to return early with `RedactedContent` instead of failing to deserialize itself according
```
```suggestion
    // to return early with `RedactedContent` instead of failing to deserialize according
```

jplatte commented on ruma-events/src/lib.rs:
```diff
@@ -255,6 +266,10 @@ pub struct EventDeHelper {
+
+    /// If this `UnsignedData` contains a redacted_because key the event is
+    /// immediately deserialized as a redacted event.
+    pub unsigned: Option<UnsignedData>,
```
We should use a separate type here, sth like `struct UnsignedDeHelper { redacted_because: Option<IgnoredAny> }`.

jplatte commented on ruma-events/src/lib.rs:
```diff
@@ -220,6 +224,13 @@ pub trait EventContent: Sized + Serialize {
+    /// If called for anything but "empty" redacted content this will error.
+    fn redacted(_event_type: &str) -> Result<Self, serde_json::Error> {
+        Err(serde::de::Error::custom("this event is not redacted"))
+    }
```
This solution with the default implementation is quite interesting, however I think we should just have more traits. Yeah, we already have lots of them, but the reason for that is that we don't want to allow invalid combinations of generic event type & event content type. So, unless I'm missing something, we need a `Redacted` version of every existing `EventContent` trait, with the same supertrait hierarchy plus `RedactedEventContent` having a supertrait of `EventContent`. This method, and `has_fields` I proposed earlier (or something equivalent) would then move to `RedactedEventContent`.

jplatte commented:
I think I mentioned this before, but the (currently only) case I actually think we should have `custom_redacted` for (although it could be solved differently too) is `m.room.aliases`: its redact function actually needs a parameter for whether to redact the content or not. Thinking about this more, this also means the associated `has_fields` function I thought of doesn't work! We actually need a method, taking `&self`, to determine whether to write the `content` field. That shouldn't make things any more complicated, though.

DevinR528 commented:
We also need `redact` methods on the `Any*Event` enums and something to deal with the fact that `redacted_because` is only a `RedactedEvent` so doesn't allow stub or stripped events. I tried out what we have so far in matrix-sdk and these were the only awkward bits.

> I actually think we should have custom_redacted for (although it could be solved differently too) is m.room.aliases

You have said that before but a good reminder, I was thinking that implementations could redact their own made-up events but this can't actually happen can it (a server has no way to know what a client is up to and vice-versa)? If the only reason we have custom redacted events is aliases should we maybe remove the custom and add a `RedactedV1AliasesEventContent` or something like that? Do we need to have a type for the old redacted aliases event, if we got sent the JSON for an old aliases event it would (de)serialize correctly? Having the catch all and semitry is nice.

I'll start working on the `RedactedEventContent` and see if that opens any posibillities that would make custom easier to implement.

jplatte commented:
> We also need `redact` methods on the `Any*Event` enums

This PR is already quite big, let's do this separately.



> and something to deal with the fact that `redacted_because` is only a `RedactedEvent` so doesn't allow stub or stripped events

I don't understand what you mean, can you expand on that?

> I was thinking that implementations could redact their own made-up events but this can't actually happen can it (a server has no way to know what a client is up to and vice-versa)?

Again not sure what you mean, sorry 😅 

> If the only reason we have custom redacted events is aliases should we maybe remove the custom and add a `RedactedV1AliasesEventContent` or something like that? Do we need to have a type for the old redacted aliases event, if we got sent the JSON for an old aliases event it would (de)serialize correctly? Having the catch all and semitry is nice.

I don't fully understand what your idea is here, but if you want, try it out.

DevinR528 commented:
>and something to deal with the fact that redacted_because is only a RedactedEvent so doesn't allow stub or stripped events

According to spec the redaction event should be inserted into the `redacted_because` field of `unsigned`, what I was getting at is that we should have a way (an enum maybe see example below) that allows this to be a stub or full event (are there stripped redaction event?)

```rust
enum AnyRedactionEvent {
    Full(RedactionEvent),
    Stub(RedactionEventStub),
    // stripped if needed
}
// then
struct UnsignedData {
    redacted_because: Option<AnyRedactionEvent>,
}
```

>I was thinking that implementations could redact their own made-up events but this can't actually happen can it (a server has no way to know what a client is up to and vice-versa)?

I see now you weren't misunderstanding I read your comment completely wrong you were just talking about the attribute not custom events in general. So disregard that paragraph.

jplatte commented:
> According to spec the redaction event should be inserted into the `redacted_because` field of `unsigned`, what I was getting at is that we should have a way (an enum maybe see example below) that allows this to be a stub or full event (are there stripped redaction event?)

I see! I guess the spec could be clearer about this. It definitely makes sense in my mind that `redacted_because` could itself contain a redacted (redaction) event. I'm not so sure about the others though. I will ask on #matrix-spec:matrix.org.

DevinR528 committed:
Add RedactedEventContent trait and super traits to match EventContent

The *EventContent derive now generates the RedactedEventContent
and EventContent trait implementations.

DevinR528 committed:
Fix trailing whitespace cargo fmt error

jplatte commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -151,6 +156,16 @@ pub fn expand_event_content(input: &DeriveInput) -> syn::Result<TokenStream> {
+                    #has_fields
+                }
+
+                fn has_deserialize_fields() -> bool {
+                    #has_fields
+                }
```
This is not called anywhere, is it? I don't think we actually need it. For redacted aliases events, I don't think there is any harm in unconditionally deserializing the content, and we can't actually know what room version the room that event is in has anyway.

DevinR528 replied:
The reason that needs to be called is `RedactedCustomEventContent` it fails to deserialize with a missing `content` field error.

jplatte replied:
Okay. Have you actually replaced the `mem::size_of` checks with this though? I couldn't find that.

DevinR528 replied:
Nope, I definitely should as now they aren't needed.

jplatte commented on ruma-events/src/lib.rs:
```diff
@@ -234,19 +260,27 @@ pub trait EventContent: Sized + Serialize {
 
 /// Marker trait for the content of a basic event.
-pub trait BasicEventContent: EventContent {}
+pub trait RedactedBasicEventContent: RedactedEventContent {}
```
These don't actually need to exit AFAIK. Only things inside rooms can be redacted, right?

jplatte replied:
Which I guess would also mean that `RedactedRoomEventContent` is useless and can be replaced by `RedactedEventContent` everywhere it is used.

DevinR528 replied:
How do you feel about all event contents that aren't redacted have this `#[ruma_event(not_redacted)]` attribute. There is no straight forward way to check if the content should be redacted, that I can think of.

DevinR528 replied:
There's 12 event contents that would need this I think.

jplatte replied:
There is a straight-forward way: Only stuff that uses the `MessageEventContent` or `StateEventContent` derive needs all the redacted bits. Everything else can just be treated the same way as `#[ruma_event(custom_redacted)]` except that we also won't write a custom redacted struct or `redact` method.

jplatte commented on ruma-events/src/room/redaction.rs:
```diff
@@ -55,33 +55,19 @@ pub struct RedactionEventStub {
 
 impl ruma_events::MessageEventContent for RedactionEventContent {}
+
+impl RedactedRoomEventContent for RedactedRedactionEventContent {}
```
Please be consistent here. Either all `*EventContent` traits should be `use`d, or none.

jplatte commented on ruma-events/src/room/aliases.rs:
```diff
@@ -3,16 +3,113 @@
+
+/// An aliases event that has been redacted.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct RedactedV1AliasesEventContent {
```
So you have added this second redacted aliases event content struct, but it's not actually part of `AnyRedactedStateEvent`, is it?

DevinR528 replied:
Correct that's pretty pointless and should probably be it's own PR? Good to know we have all the plumbing to make it work though.

jplatte replied:
I'm still sceptical about that working. Could you implement `RedactedAliasesEventContent` as containing an `Option<Vec<RoomAliasId>>` in this PR, and then move this to its own PR?

DevinR528 replied:
Sounds good! Add the `Option<...>` then just leave it for later.

jplatte commented on ruma-events/src/lib.rs:
```diff
@@ -237,6 +241,47 @@ pub trait MessageEventContent: RoomEventContent {}
+    /// Constructs the redacted event content.
+    ///
+    /// If called for anything but "empty" redacted content this will error.
+    fn redacted(_event_type: &str) -> Result<Self, serde_json::Error> {
```
I wanted to suggest renaming this to `empty` and having it as the first function in the trait two times now. Somehow I failed both times, due to not finding the right place to put the annotation in the diff and then forgetting. Oh well, third time lucky :D

DevinR528 committed:
Add emit_redacted flag to all expand_*_event_content fns

Rename RedactedEventContent::redacted -> empty

DevinR528 commented:
Ah sorry about that I squashed a commit the wrong way, I should have squashed the last one into the one I ended up squashing.
```rust
oldest commit
middle // I squashed this into the commit above it
new // I should have squashed this and rewrote the commit?
```

jplatte commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -61,7 +59,7 @@ impl Parse for EventMeta {
 
 /// Create an `EventContent` implementation for a struct.
-pub fn expand_event_content(input: &DeriveInput) -> syn::Result<TokenStream> {
+pub fn expand_event_content(input: &DeriveInput, emit_redacted: bool) -> syn::Result<TokenStream> {
```
This is actually the only function that needs this parameter. Every other `_expand_event_content` is only ever called with `true`, or `false`.

jplatte reviewed (commented):
So the only missing thing now is the `aliases` field for `RedactedAliasesEventContent` and corresponding `RedactedEventContent` implementation changes, right?

DevinR528 committed:
Make aliases redacted event content optional for version 1 compat

jplatte commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -7,39 +7,181 @@ use syn::{
+    // Do not emit any redacted event code.
+    syn::custom_keyword!(custom_redacted);
+    // This event is not redact-able
+    syn::custom_keyword!(not_redacted);
```
```suggestion
```

jplatte commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -7,39 +7,181 @@ use syn::{
+            #[derive(Clone, Debug, ::serde::Deserialize, ::serde::Serialize)]
+            pub struct #redacted_ident #redacted_fields
+
+            impl ::ruma_events::EventContent for #redacted_ident {
```
Is this impl generation not the same as the existing? Unless there's a significant difference, this part of the codegen should not be duplicated. I'd be okay with refactoring in a followup PR though, if you want.

jplatte commented on ruma-events-macros/src/event_content.rs:
```diff
@@ -101,25 +245,56 @@ pub fn expand_room_event_content(input: DeriveInput) -> syn::Result<TokenStream>
     let room_ev_content = expand_room_event_content(input)?;
 
+    let redacted_marker_trait = if needs_redacted(input) {
+        let ident = quote::format_ident!("Redacted{}", &ident);
```
I think this is used often enough to deserve a `use` at the start of the file.

DevinR528 replied:
I've started a seperate PR that converts all of ruma-events to use `format_ident` and I fix this in that PR, if that cool.

jplatte replied:
Sure.

jplatte commented on ruma-events/src/room/redaction.rs:
```diff
@@ -64,84 +67,11 @@ pub struct RedactionEventContent {
+
+impl MessageEventContent for RedactionEventContent {}
+
+// A redacted custom event must satisfy both state and message events.
```
We don't have this kind of comment for the `impl`s above. Also I'm not sure about the wording. If we have this at all, it should say something like `can be used as either … or …` or `can be used both as … and …`.

jplatte commented on ruma-events/src/room/member.rs:
```diff
@@ -54,6 +54,7 @@ pub struct MemberEventContent {
     pub is_direct: Option<bool>,
 
     /// The membership state of this user.
+    #[ruma_event(skip_redacted)]
```
Why is this not producing an error?

```suggestion
    #[ruma_event(skip_redaction)]
```

jplatte commented on ruma-events/src/event_kinds.rs:
```diff
@@ -149,6 +191,76 @@ pub struct StrippedStateEventStub<C: StateEventContent> {
+
+/// A stripped-down redacted state event.
+#[derive(Clone, Debug, Event)]
+pub struct RedactedStrippedStateEventStub<C: RedactedStateEventContent> {
```
No `RedactedStrippedStateEvent`?

DevinR528 replied:
Is there supposed to also be `StrippedStateEvent`? If so there is no regular event struct for this either.

jplatte replied:
Oh, right. 😄 

Yeah no, stripped state events are only ever received as part of the sync response ^^

DevinR528 replied:
So I don't need to change anything right?

jplatte replied:
No ^^

jplatte commented on ruma-events/src/enums.rs:
```diff
@@ -117,6 +125,10 @@ pub enum AnyRoomEventStub {
+    /// Any state event stub that has been redacted.
+    RedactedState(AnyRedactedStateEventStub),
 }
 
```
Maybe something for a followup PR too, but we do now need enums over `AnyStateEvent | AnyRedactedStateEvent`, `AnyStateEventStub | AnyRedactedStateEventStub` and `AnyStrippedStateEventStub | AnyRedactedStrippedStateEventStub`. Maybe we actually need to use a `Regular` prefix for the generated `Message` and `State` enums and reuse their previous named for these new enums.

DevinR528 replied:
I think the rename makes sense. Probably be best to do it as a new PR, and maybe after the `format_ident` and fixing of the brittle enum name generation (fixing this stuff `ident.to_string().replace...` and making the `event_enum!` macro only take certain `name/kind: Message`)?

jplatte replied:
Sounds good.

jplatte commented on ruma-events/src/custom.rs:
```diff
@@ -42,3 +50,45 @@ impl EphemeralRoomEventContent for CustomEventContent {}
+    }
+}
+
+// A redacted custom event must satisfy both state and message events.
```
Same thing as for the redaction event.

jplatte reviewed (commented):
So I did another full review and found a few more issues 😅

DevinR528 committed:
Emit error when ruma_event field attr is not skip_redaction

DevinR528 committed:
Revert EventMeta parsing logic
